src/software/new_geom/polynomial.cpp

```diff
@@ -0,0 +1,128 @@
+#include "software/new_geom/polynomial.h"
+
+Polynomial::Polynomial() {}
+
+Polynomial::Polynomial(const std::vector<double> &coeffs) : coeffs(coeffs) {}
+
+Polynomial::Polynomial(const std::initializer_list<double> &coeffs) : coeffs(coeffs) {}
+
+Polynomial::Polynomial(const std::pair<double, double> &constraint1,
+                       const std::pair<double, double> &constraint2)
+{
+    if (constraint1.first == constraint2.first)
+    {
+        throw std::invalid_argument("Both inputs are equal - does not define a function");
+    }
+    double slope = (constraint2.second - constraint1.second) /
+                   (constraint2.first - constraint1.first);
+    coeffs.push_back(constraint1.second - (constraint1.first * slope));
+    coeffs.push_back(slope);
+}
+
+double Polynomial::getCoeff(unsigned int order) const
+{
+    return order >= coeffs.size() ? 0 : coeffs[order];
+}
+
+void Polynomial::setCoeff(unsigned int order, double coeff)
+{
+    if (order >= coeffs.size())
+    {
+        coeffs.resize(order + 1, 0);
+    }
+    coeffs[order] = coeff;
+}
+
+unsigned int Polynomial::getOrder() const
+{
+    if (coeffs.size() != 0)
+    {
+        for (size_t i = coeffs.size() - 1; i >= 0; i++)
+        {
+            if (std::abs(coeffs[i]) >= EPSILON)
+            {
+                return i;
+            }
+        }
+    }
+    // Zero polynomial treated as an order zero polynomial
+    return 0;
+}
+
+double Polynomial::valueAt(double val) const
+{
+    // If this has a large performance impact, could be improved
+    // by using Horner's Method:
+    // https://www.geeksforgeeks.org/horners-method-polynomial-evaluation/
+    double retval = 0;
+    for (unsigned int i = 0; i <= getOrder(); i++)
```

Comment:
I see, you check order here. If store this outside the loop (so we don't call this every loop), then I think it's OK.

Reply:
Addressed in 37e8554cd0bbfba2772160f2707effbccccc